haystack/preview/components/builders/answers_builder.py

```diff
@@ -0,0 +1,153 @@
+        reference_pattern = reference_pattern or self.reference_pattern
+
+        all_answers = []
+        for query, reply_list, doc_list, meta_list in zip_longest(queries, replies, documents, metadata, fillvalue=[]):
+            extracted_answer_strings = AnswersBuilder._extract_answer_strings(reply_list, pattern)
+            if doc_list and reference_pattern:
+                reference_idxs = AnswersBuilder._extract_reference_idxs(reply_list, reference_pattern)
+            else:
+                reference_idxs = [[doc_idx for doc_idx, _ in enumerate(doc_list)] * len(reply_list)]
+
+            answers_for_cur_query = []
+            for answer_string, doc_idxs, meta in zip_longest(extracted_answer_strings, reference_idxs, meta_list):
+                referenced_docs = []
+                if doc_idxs:
+                    referenced_docs = [doc_list[idx] for idx in doc_idxs if idx < len(doc_list)]
+
+                answer = GeneratedAnswer(data=answer_string, query=query, documents=referenced_docs, metadata=meta)
+                answers_for_cur_query.append(answer)
+
+            all_answers.append(answers_for_cur_query)
+
+        return all_answers
+
+    def to_dict(self) -> Dict[str, Any]:
+        """
+        Serialize this component to a dictionary.
+        """
+        return default_to_dict(self, pattern=self.pattern, reference_pattern=self.reference_pattern)
+
+    @classmethod
+    def from_dict(cls, data: Dict[str, Any]) -> "AnswersBuilder":
+        """
+        Deserialize this component from a dictionary.
+        """
+        return default_from_dict(cls, data)
+
+    @staticmethod
+    def _extract_answer_strings(replies: List[str], pattern: Optional[str] = None) -> List[str]:
+        """
+        Extract the answer strings from the generator output using the specified pattern.
+        If no pattern is specified, the whole string is used as the answer.
+
+        :param replies: The output of the Generator. A list of strings.
+        :param pattern: The regular expression pattern to use to extract the answer text from the generator output.
+        """
+        if pattern is None:
+            return replies
+
+        extracted_answers = []
+        for reply in replies:
+            if match := re.search(pattern, reply):
+                # No capture group in pattern -> use the whole match as answer
+                if not match.lastindex:
+                    extracted_answers.append(match.group(0))
+                # One capture group in pattern -> use the capture group as answer
+                elif match.lastindex == 1:
+                    extracted_answers.append(match.group(1))
+                # Multiple capture groups in pattern -> raise error
+                else:
+                    raise ValueError(
+                        f"Pattern '{pattern}' contains multiple capture groups. "
+                        f"Please specify a pattern with at most one capture group."
+                    )
```

Comment:
I'd rather not fail at runtime this late... Is there a way to figure if the regex contains more than one capture group in init (or at the start of `run()` if given at runtime)? if not, I'd raise a WARN log in here and just use the first capture group. Not nice but better than failing imho